{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## In this notebook I use Bowtie to align geoduck RNASeq files to my >70k scaffolds. \n",
    "\n",
    "### The Bowtie [user manual](http://bowtie-bio.sourceforge.net/tutorial.shtml) has a great tutorial, which is what I referenced to perform the processes in this notebook. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/shlaura3/Documents/SAFS/FISH_546_Bioinformatics/546-Bioinformatics/2016-10_Geo-Ann-Project/Jupyter-Notebooks'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# I installed Bowtie via the following command: \n",
    "conda install -c bioconda bowtie=1.1.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No index, query, or output file specified!\r\n",
      "Usage: \r\n",
      "bowtie [options]* <ebwt> {-1 <m1> -2 <m2> | --12 <r> | <s>} [<hit>]\r\n",
      "\r\n",
      "  <m1>    Comma-separated list of files containing upstream mates (or the\r\n",
      "          sequences themselves, if -c is set) paired with mates in <m2>\r\n",
      "  <m2>    Comma-separated list of files containing downstream mates (or the\r\n",
      "          sequences themselves if -c is set) paired with mates in <m1>\r\n",
      "  <r>     Comma-separated list of files containing Crossbow-style reads.  Can be\r\n",
      "          a mixture of paired and unpaired.  Specify \"-\" for stdin.\r\n",
      "  <s>     Comma-separated list of files containing unpaired reads, or the\r\n",
      "          sequences themselves, if -c is set.  Specify \"-\" for stdin.\r\n",
      "  <hit>   File to write hits to (default: stdout)\r\n",
      "Input:\r\n",
      "  -q                 query input files are FASTQ .fq/.fastq (default)\r\n",
      "  -f                 query input files are (multi-)FASTA .fa/.mfa\r\n",
      "  -r                 query input files are raw one-sequence-per-line\r\n",
      "  -c                 query sequences given on cmd line (as <mates>, <singles>)\r\n",
      "  -C                 reads and index are in colorspace\r\n",
      "  -Q/--quals <file>  QV file(s) corresponding to CSFASTA inputs; use with -f -C\r\n",
      "  --Q1/--Q2 <file>   same as -Q, but for mate files 1 and 2 respectively\r\n",
      "  -s/--skip <int>    skip the first <int> reads/pairs in the input\r\n",
      "  -u/--qupto <int>   stop after first <int> reads/pairs (excl. skipped reads)\r\n",
      "  -5/--trim5 <int>   trim <int> bases from 5' (left) end of reads\r\n",
      "  -3/--trim3 <int>   trim <int> bases from 3' (right) end of reads\r\n",
      "  --phred33-quals    input quals are Phred+33 (default)\r\n",
      "  --phred64-quals    input quals are Phred+64 (same as --solexa1.3-quals)\r\n",
      "  --solexa-quals     input quals are from GA Pipeline ver. < 1.3\r\n",
      "  --solexa1.3-quals  input quals are from GA Pipeline ver. >= 1.3\r\n",
      "  --integer-quals    qualities are given as space-separated integers (not ASCII)\r\n",
      "  --large-index      force usage of a 'large' index, even if a small one is present\r\n",
      "Alignment:\r\n",
      "  -v <int>           report end-to-end hits w/ <=v mismatches; ignore qualities\r\n",
      "    or\r\n",
      "  -n/--seedmms <int> max mismatches in seed (can be 0-3, default: -n 2)\r\n",
      "  -e/--maqerr <int>  max sum of mismatch quals across alignment for -n (def: 70)\r\n",
      "  -l/--seedlen <int> seed length for -n (default: 28)\r\n",
      "  --nomaqround       disable Maq-like quality rounding for -n (nearest 10 <= 30)\r\n",
      "  -I/--minins <int>  minimum insert size for paired-end alignment (default: 0)\r\n",
      "  -X/--maxins <int>  maximum insert size for paired-end alignment (default: 250)\r\n",
      "  --fr/--rf/--ff     -1, -2 mates align fw/rev, rev/fw, fw/fw (default: --fr)\r\n",
      "  --nofw/--norc      do not align to forward/reverse-complement reference strand\r\n",
      "  --maxbts <int>     max # backtracks for -n 2/3 (default: 125, 800 for --best)\r\n",
      "  --pairtries <int>  max # attempts to find mate for anchor hit (default: 100)\r\n",
      "  -y/--tryhard       try hard to find valid alignments, at the expense of speed\r\n",
      "  --chunkmbs <int>   max megabytes of RAM for best-first search frames (def: 64)\r\n",
      "Reporting:\r\n",
      "  -k <int>           report up to <int> good alignments per read (default: 1)\r\n",
      "  -a/--all           report all alignments per read (much slower than low -k)\r\n",
      "  -m <int>           suppress all alignments if > <int> exist (def: no limit)\r\n",
      "  -M <int>           like -m, but reports 1 random hit (MAPQ=0); requires --best\r\n",
      "  --best             hits guaranteed best stratum; ties broken by quality\r\n",
      "  --strata           hits in sub-optimal strata aren't reported (requires --best)\r\n",
      "Output:\r\n",
      "  -t/--time          print wall-clock time taken by search phases\r\n",
      "  -B/--offbase <int> leftmost ref offset = <int> in bowtie output (default: 0)\r\n",
      "  --quiet            print nothing but the alignments\r\n",
      "  --refout           write alignments to files refXXXXX.map, 1 map per reference\r\n",
      "  --refidx           refer to ref. seqs by 0-based index rather than name\r\n",
      "  --al <fname>       write aligned reads/pairs to file(s) <fname>\r\n",
      "  --un <fname>       write unaligned reads/pairs to file(s) <fname>\r\n",
      "  --max <fname>      write reads/pairs over -m limit to file(s) <fname>\r\n",
      "  --suppress <cols>  suppresses given columns (comma-delim'ed) in default output\r\n",
      "  --fullref          write entire ref name (default: only up to 1st space)\r\n",
      "Colorspace:\r\n",
      "  --snpphred <int>   Phred penalty for SNP when decoding colorspace (def: 30)\r\n",
      "     or\r\n",
      "  --snpfrac <dec>    approx. fraction of SNP bases (e.g. 0.001); sets --snpphred\r\n",
      "  --col-cseq         print aligned colorspace seqs as colors, not decoded bases\r\n",
      "  --col-cqual        print original colorspace quals, not decoded quals\r\n",
      "  --col-keepends     keep nucleotides at extreme ends of decoded alignment\r\n",
      "SAM:\r\n",
      "  -S/--sam           write hits in SAM format\r\n",
      "  --mapq <int>       default mapping quality (MAPQ) to print for SAM alignments\r\n",
      "  --sam-nohead       supppress header lines (starting with @) for SAM output\r\n",
      "  --sam-nosq         supppress @SQ header lines for SAM output\r\n",
      "  --sam-RG <text>    add <text> (usually \"lab=value\") to @RG line of SAM header\r\n",
      "Performance:\r\n",
      "  -o/--offrate <int> override offrate of index; must be >= index's offrate\r\n",
      "  -p/--threads <int> number of alignment threads to launch (default: 1)\r\n",
      "  --mm               use memory-mapped I/O for index; many 'bowtie's can share\r\n",
      "  --shmem            use shared mem for index; many 'bowtie's can share\r\n",
      "Other:\r\n",
      "  --seed <int>       seed for random number generator\r\n",
      "  --verbose          verbose output (for debugging)\r\n",
      "  --version          print version information and quit\r\n",
      "  -h/--help          print this usage message\r\n"
     ]
    }
   ],
   "source": [
    "# Bowtie was automatically installed in the `anaconda` bin:\n",
    "! /Users/shlaura3/anaconda/bin/bowtie"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: \r\n",
      "bowtie [options]* <ebwt> {-1 <m1> -2 <m2> | --12 <r> | <s>} [<hit>]\r\n",
      "\r\n",
      "  <m1>    Comma-separated list of files containing upstream mates (or the\r\n",
      "          sequences themselves, if -c is set) paired with mates in <m2>\r\n",
      "  <m2>    Comma-separated list of files containing downstream mates (or the\r\n",
      "          sequences themselves if -c is set) paired with mates in <m1>\r\n",
      "  <r>     Comma-separated list of files containing Crossbow-style reads.  Can be\r\n",
      "          a mixture of paired and unpaired.  Specify \"-\" for stdin.\r\n",
      "  <s>     Comma-separated list of files containing unpaired reads, or the\r\n",
      "          sequences themselves, if -c is set.  Specify \"-\" for stdin.\r\n",
      "  <hit>   File to write hits to (default: stdout)\r\n",
      "Input:\r\n",
      "  -q                 query input files are FASTQ .fq/.fastq (default)\r\n",
      "  -f                 query input files are (multi-)FASTA .fa/.mfa\r\n",
      "  -r                 query input files are raw one-sequence-per-line\r\n",
      "  -c                 query sequences given on cmd line (as <mates>, <singles>)\r\n",
      "  -C                 reads and index are in colorspace\r\n",
      "  -Q/--quals <file>  QV file(s) corresponding to CSFASTA inputs; use with -f -C\r\n",
      "  --Q1/--Q2 <file>   same as -Q, but for mate files 1 and 2 respectively\r\n",
      "  -s/--skip <int>    skip the first <int> reads/pairs in the input\r\n",
      "  -u/--qupto <int>   stop after first <int> reads/pairs (excl. skipped reads)\r\n",
      "  -5/--trim5 <int>   trim <int> bases from 5' (left) end of reads\r\n",
      "  -3/--trim3 <int>   trim <int> bases from 3' (right) end of reads\r\n",
      "  --phred33-quals    input quals are Phred+33 (default)\r\n",
      "  --phred64-quals    input quals are Phred+64 (same as --solexa1.3-quals)\r\n",
      "  --solexa-quals     input quals are from GA Pipeline ver. < 1.3\r\n",
      "  --solexa1.3-quals  input quals are from GA Pipeline ver. >= 1.3\r\n",
      "  --integer-quals    qualities are given as space-separated integers (not ASCII)\r\n",
      "  --large-index      force usage of a 'large' index, even if a small one is present\r\n",
      "Alignment:\r\n",
      "  -v <int>           report end-to-end hits w/ <=v mismatches; ignore qualities\r\n",
      "    or\r\n",
      "  -n/--seedmms <int> max mismatches in seed (can be 0-3, default: -n 2)\r\n",
      "  -e/--maqerr <int>  max sum of mismatch quals across alignment for -n (def: 70)\r\n",
      "  -l/--seedlen <int> seed length for -n (default: 28)\r\n",
      "  --nomaqround       disable Maq-like quality rounding for -n (nearest 10 <= 30)\r\n",
      "  -I/--minins <int>  minimum insert size for paired-end alignment (default: 0)\r\n",
      "  -X/--maxins <int>  maximum insert size for paired-end alignment (default: 250)\r\n",
      "  --fr/--rf/--ff     -1, -2 mates align fw/rev, rev/fw, fw/fw (default: --fr)\r\n",
      "  --nofw/--norc      do not align to forward/reverse-complement reference strand\r\n",
      "  --maxbts <int>     max # backtracks for -n 2/3 (default: 125, 800 for --best)\r\n",
      "  --pairtries <int>  max # attempts to find mate for anchor hit (default: 100)\r\n",
      "  -y/--tryhard       try hard to find valid alignments, at the expense of speed\r\n",
      "  --chunkmbs <int>   max megabytes of RAM for best-first search frames (def: 64)\r\n",
      "Reporting:\r\n",
      "  -k <int>           report up to <int> good alignments per read (default: 1)\r\n",
      "  -a/--all           report all alignments per read (much slower than low -k)\r\n",
      "  -m <int>           suppress all alignments if > <int> exist (def: no limit)\r\n",
      "  -M <int>           like -m, but reports 1 random hit (MAPQ=0); requires --best\r\n",
      "  --best             hits guaranteed best stratum; ties broken by quality\r\n",
      "  --strata           hits in sub-optimal strata aren't reported (requires --best)\r\n",
      "Output:\r\n",
      "  -t/--time          print wall-clock time taken by search phases\r\n",
      "  -B/--offbase <int> leftmost ref offset = <int> in bowtie output (default: 0)\r\n",
      "  --quiet            print nothing but the alignments\r\n",
      "  --refout           write alignments to files refXXXXX.map, 1 map per reference\r\n",
      "  --refidx           refer to ref. seqs by 0-based index rather than name\r\n",
      "  --al <fname>       write aligned reads/pairs to file(s) <fname>\r\n",
      "  --un <fname>       write unaligned reads/pairs to file(s) <fname>\r\n",
      "  --max <fname>      write reads/pairs over -m limit to file(s) <fname>\r\n",
      "  --suppress <cols>  suppresses given columns (comma-delim'ed) in default output\r\n",
      "  --fullref          write entire ref name (default: only up to 1st space)\r\n",
      "Colorspace:\r\n",
      "  --snpphred <int>   Phred penalty for SNP when decoding colorspace (def: 30)\r\n",
      "     or\r\n",
      "  --snpfrac <dec>    approx. fraction of SNP bases (e.g. 0.001); sets --snpphred\r\n",
      "  --col-cseq         print aligned colorspace seqs as colors, not decoded bases\r\n",
      "  --col-cqual        print original colorspace quals, not decoded quals\r\n",
      "  --col-keepends     keep nucleotides at extreme ends of decoded alignment\r\n",
      "SAM:\r\n",
      "  -S/--sam           write hits in SAM format\r\n",
      "  --mapq <int>       default mapping quality (MAPQ) to print for SAM alignments\r\n",
      "  --sam-nohead       supppress header lines (starting with @) for SAM output\r\n",
      "  --sam-nosq         supppress @SQ header lines for SAM output\r\n",
      "  --sam-RG <text>    add <text> (usually \"lab=value\") to @RG line of SAM header\r\n",
      "Performance:\r\n",
      "  -o/--offrate <int> override offrate of index; must be >= index's offrate\r\n",
      "  -p/--threads <int> number of alignment threads to launch (default: 1)\r\n",
      "  --mm               use memory-mapped I/O for index; many 'bowtie's can share\r\n",
      "  --shmem            use shared mem for index; many 'bowtie's can share\r\n",
      "Other:\r\n",
      "  --seed <int>       seed for random number generator\r\n",
      "  --verbose          verbose output (for debugging)\r\n",
      "  --version          print version information and quit\r\n",
      "  -h/--help          print this usage message\r\n"
     ]
    }
   ],
   "source": [
    "! /Users/shlaura3/anaconda/bin/bowtie -help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"../data/Panopea_generosa_scaff-70k_bowtie-index.*.ebwt\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 5 (one in 32)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  ../data/Panopea_generosa_scaff-70k.scafSeq\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 451843\n",
      "Using parameters --bmax 338883 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 338883 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "  Binary sorting into buckets\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Binary sorting into buckets time: 00:00:00\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 7; iterating...\n",
      "  Binary sorting into buckets\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Binary sorting into buckets time: 00:00:00\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 0; iterating...\n",
      "  Binary sorting into buckets\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Binary sorting into buckets time: 00:00:00\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 258195 (target: 338882)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 320064\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 320065\n",
      "Getting block 2 of 7\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 128366\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 128367\n",
      "Getting block 3 of 7\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 323687\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 323688\n",
      "Getting block 4 of 7\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 284762\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 284763\n",
      "Getting block 5 of 7\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 208405\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 208406\n",
      "Getting block 6 of 7\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 324235\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 324236\n",
      "Getting block 7 of 7\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 217848\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 217849\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 605139\n",
      "fchr[G]: 900679\n",
      "fchr[T]: 1194923\n",
      "fchr[$]: 1807373\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4718825 bytes to primary EBWT file: ../data/Panopea_generosa_scaff-70k_bowtie-index.1.ebwt\n",
      "Wrote 225928 bytes to secondary EBWT file: ../data/Panopea_generosa_scaff-70k_bowtie-index.2.ebwt\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 1807373\n",
      "    bwtLen: 1807374\n",
      "    sz: 451844\n",
      "    bwtSz: 451844\n",
      "    lineRate: 6\n",
      "    linesPerSide: 1\n",
      "    offRate: 5\n",
      "    offMask: 0xffffffe0\n",
      "    isaRate: -1\n",
      "    isaMask: 0xffffffff\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 56481\n",
      "    offsSz: 225924\n",
      "    isaLen: 0\n",
      "    isaSz: 0\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 56\n",
      "    sideBwtLen: 224\n",
      "    numSidePairs: 4035\n",
      "    numSides: 8070\n",
      "    numLines: 8070\n",
      "    ebwtTotLen: 516480\n",
      "    ebwtTotSz: 516480\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:01\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 451843\n",
      "Using parameters --bmax 338883 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 338883 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "  Binary sorting into buckets\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Binary sorting into buckets time: 00:00:00\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 7; iterating...\n",
      "  Binary sorting into buckets\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Binary sorting into buckets time: 00:00:00\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 0; iterating...\n",
      "  Binary sorting into buckets\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Binary sorting into buckets time: 00:00:01\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 225921 (target: 338882)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 185024\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 185025\n",
      "Getting block 2 of 8\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 307323\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 307324\n",
      "Getting block 3 of 8\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 235741\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 235742\n",
      "Getting block 4 of 8\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 256476\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 256477\n",
      "Getting block 5 of 8\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 194629\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 194630\n",
      "Getting block 6 of 8\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 220626\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 220627\n",
      "Getting block 7 of 8\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 215802\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 215803\n",
      "Getting block 8 of 8\n",
      "  Reserving size (338883) for bucket\n",
      "  Calculating Z arrays\n",
      "  Calculating Z arrays time: 00:00:00\n",
      "  Entering block accumulator loop:\n",
      "  10%\n",
      "  20%\n",
      "  30%\n",
      "  40%\n",
      "  50%\n",
      "  60%\n",
      "  70%\n",
      "  80%\n",
      "  90%\n",
      "  100%\n",
      "  Block accumulator loop time: 00:00:00\n",
      "  Sorting block of length 191745\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 191746\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 605139\n",
      "fchr[G]: 900679\n",
      "fchr[T]: 1194923\n",
      "fchr[$]: 1807373\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4718825 bytes to primary EBWT file: ../data/Panopea_generosa_scaff-70k_bowtie-index.rev.1.ebwt\n",
      "Wrote 225928 bytes to secondary EBWT file: ../data/Panopea_generosa_scaff-70k_bowtie-index.rev.2.ebwt\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 1807373\n",
      "    bwtLen: 1807374\n",
      "    sz: 451844\n",
      "    bwtSz: 451844\n",
      "    lineRate: 6\n",
      "    linesPerSide: 1\n",
      "    offRate: 5\n",
      "    offMask: 0xffffffe0\n",
      "    isaRate: -1\n",
      "    isaMask: 0xffffffff\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 56481\n",
      "    offsSz: 225924\n",
      "    isaLen: 0\n",
      "    isaSz: 0\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 56\n",
      "    sideBwtLen: 224\n",
      "    numSidePairs: 4035\n",
      "    numSides: 8070\n",
      "    numLines: 8070\n",
      "    ebwtTotLen: 516480\n",
      "    ebwtTotSz: 516480\n",
      "    reverse: 0\n",
      "Total time for backward call to driver() for mirror index: 00:00:01\n"
     ]
    }
   ],
   "source": [
    "# Now I'll build my \"index\" (aka database) out of my >70k genome scaffolds;\n",
    "# see line 1429 in the [bowtie manual](https://github.com/BenLangmead/bowtie/blob/master/MANUAL) for the bowtie-build options\n",
    "!/Users/shlaura3/anaconda/bin/bowtie-build \\\n",
    "../data/Panopea_generosa_scaff-70k.scafSeq \\\n",
    "../data/Panopea_generosa_scaff-70k_bowtie-index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geo-70k-scaff-annotations-merged-table.tabular\r\n",
      "Geo-v3-join-uniprot-all0916-condensed.tab\r\n",
      "Geoduck-transcriptome-v3.fa.zip\r\n",
      "Geoduck-transcriptome-v3.fa_db.nhr\r\n",
      "Geoduck-transcriptome-v3.fa_db.nin\r\n",
      "Geoduck-transcriptome-v3.fa_db.nsq\r\n",
      "\u001b[31mGeoduck-transcriptome-v3annotated.fa\u001b[m\u001b[m*\r\n",
      "Panopea_generosa_No-Line-Breaks.scafSeq.fai\r\n",
      "Panopea_generosa_ScafSeq.genome\r\n",
      "Panopea_generosa_scaff-70k.scafSeq\r\n",
      "Panopea_generosa_scaff-70k.scafSeq.fai\r\n",
      "Panopea_generosa_scaff-70k_bowtie-index.1.ebwt\r\n",
      "Panopea_generosa_scaff-70k_bowtie-index.2.ebwt\r\n",
      "Panopea_generosa_scaff-70k_bowtie-index.3.ebwt\r\n",
      "Panopea_generosa_scaff-70k_bowtie-index.4.ebwt\r\n",
      "Panopea_generosa_scaff-70k_bowtie-index.rev.1.ebwt\r\n",
      "Panopea_generosa_scaff-70k_bowtie-index.rev.2.ebwt\r\n",
      "Panopea_generosa_scaff-70k_db.nhr\r\n",
      "Panopea_generosa_scaff-70k_db.nin\r\n",
      "Panopea_generosa_scaff-70k_db.nsq\r\n",
      "Phel_countdata.txt\r\n",
      "Phel_transcriptome.fasta\r\n",
      "hairpin.fa\r\n"
     ]
    }
   ],
   "source": [
    "ls ../data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/shlaura3/Documents/SAFS/FISH_546_Bioinformatics/546-Bioinformatics/2016-10_Geo-Ann-Project/data\n"
     ]
    }
   ],
   "source": [
    "cd ../data/ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 24.9G  100 24.9G    0     0   857k      0  8:28:43  8:28:43 --:--:--  916k\n",
      " 29 25.0G   29 7478M    0     0   912k      0  7:59:11  2:19:55  5:39:16  914k"
     ]
    }
   ],
   "source": [
    "# download geoduck RNASeq files, which are paired-end reads, samples from gonads in 1 male & 1 female\n",
    "! curl -O http://owl.fish.washington.edu/halfshell/bu-data-genomic/tentacle/Geoduck_v3/Geo_Pool_F_GGCTAC_L006_R2_001_val_1.fq \\\n",
    "-O http://owl.fish.washington.edu/halfshell/bu-data-genomic/tentacle/Geoduck_v3/Geo_Pool_F_GGCTAC_L006_R1_001_val_2.fq \\\n",
    "-O http://owl.fish.washington.edu/halfshell/bu-data-genomic/tentacle/Geoduck_v3/Geo_Pool_M_CTTGTA_L006_R2_001_val_1.fq \\\n",
    "-O http://owl.fish.washington.edu/halfshell/bu-data-genomic/tentacle/Geoduck_v3/Geo_Pool_M_CTTGTA_L006_R1_001_val_2.fq    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geo-70k-scaff-annotations-merged-table.tabular\r\n",
      "Geo-v3-join-uniprot-all0916-condensed.tab\r\n",
      "Geo_Pool_F_GGCTAC_L006_R1_001_val_2.fq\r\n",
      "Geo_Pool_F_GGCTAC_L006_R2_001_val_1.fq\r\n",
      "Geoduck-transcriptome-v3.fa.zip\r\n",
      "Geoduck-transcriptome-v3.fa_db.nhr\r\n",
      "Geoduck-transcriptome-v3.fa_db.nin\r\n",
      "Geoduck-transcriptome-v3.fa_db.nsq\r\n",
      "\u001b[31mGeoduck-transcriptome-v3annotated.fa\u001b[m\u001b[m*\r\n",
      "Panopea_generosa_No-Line-Breaks.scafSeq.fai\r\n",
      "Panopea_generosa_ScafSeq.genome\r\n",
      "Panopea_generosa_scaff-70k.scafSeq\r\n",
      "Panopea_generosa_scaff-70k.scafSeq.fai\r\n",
      "Panopea_generosa_scaff-70k_bowtie-index.1.ebwt\r\n",
      "Panopea_generosa_scaff-70k_bowtie-index.2.ebwt\r\n",
      "Panopea_generosa_scaff-70k_bowtie-index.3.ebwt\r\n",
      "Panopea_generosa_scaff-70k_bowtie-index.4.ebwt\r\n",
      "Panopea_generosa_scaff-70k_bowtie-index.rev.1.ebwt\r\n",
      "Panopea_generosa_scaff-70k_bowtie-index.rev.2.ebwt\r\n",
      "Panopea_generosa_scaff-70k_db.nhr\r\n",
      "Panopea_generosa_scaff-70k_db.nin\r\n",
      "Panopea_generosa_scaff-70k_db.nsq\r\n",
      "Phel_countdata.txt\r\n",
      "Phel_transcriptome.fasta\r\n",
      "hairpin.fa\r\n"
     ]
    }
   ],
   "source": [
    "ls data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not locate a Bowtie index corresponding to basename \"../data/Panopea_generosascaff-70k_gonadF_.sam\"\r\n",
      "Command: bowtie --wrapper basic-0 -x ../data/Panopea_generosa_scaff-70k_bowtie-index -1 ../data/Geo_Pool_F_GGCTAC_L006_R2_001_val_1.fq -2 ../data/Geo_Pool_F_GGCTAC_L006_R1_001_val_2.fq -S ../data/Panopea_generosascaff-70k_gonadF_.sam \r\n"
     ]
    }
   ],
   "source": [
    "# Now, use bowtie to first map female gonad paired-end reads to scaffold index\n",
    "! /Users/shlaura3/anaconda/bin/bowtie \\\n",
    "-x ../data/Panopea_generosa_scaff-70k_bowtie-index \\\n",
    "-1 ../data/Geo_Pool_F_GGCTAC_L006_R2_001_val_1.fq \\\n",
    "-2 ../data/Geo_Pool_F_GGCTAC_L006_R1_001_val_2.fq \\\n",
    "-S ../data/Panopea_generosascaff-70k_gonadF_.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"../data/Panopea_generosa_scaff-70k_bowtie2-index.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  ../data/Panopea_generosa_scaff-70k.fasta\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 451843\n",
      "Using parameters --bmax 338883 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 338883 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 7; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 258195 (target: 338882)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (338883) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 299530 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 299531 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (338883) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 148900 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 148901 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (338883) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 323687 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 323688 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (338883) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 284762 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 284763 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (338883) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 208405 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 208406 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (338883) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 324235 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 324236 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (338883) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 217848 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 217849 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 605139\n",
      "fchr[G]: 900679\n",
      "fchr[T]: 1194923\n",
      "fchr[$]: 1807373\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4804841 bytes to primary EBWT file: ../data/Panopea_generosa_scaff-70k_bowtie2-index.1.bt2\n",
      "Wrote 451848 bytes to secondary EBWT file: ../data/Panopea_generosa_scaff-70k_bowtie2-index.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 1807373\n",
      "    bwtLen: 1807374\n",
      "    sz: 451844\n",
      "    bwtSz: 451844\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 112961\n",
      "    offsSz: 451844\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 9414\n",
      "    numLines: 9414\n",
      "    ebwtTotLen: 602496\n",
      "    ebwtTotSz: 602496\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:01\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:01\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 451843\n",
      "Using parameters --bmax 338883 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 338883 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 258195 (target: 338882)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (338883) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 302407 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 302408 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (338883) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 170619 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 170620 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (338883) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 287319 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 287320 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (338883) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 330746 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 330747 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (338883) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 235349 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 235350 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (338883) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 151345 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 151346 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (338883) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 329582 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 329583 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 605139\n",
      "fchr[G]: 900679\n",
      "fchr[T]: 1194923\n",
      "fchr[$]: 1807373\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4804841 bytes to primary EBWT file: ../data/Panopea_generosa_scaff-70k_bowtie2-index.rev.1.bt2\n",
      "Wrote 451848 bytes to secondary EBWT file: ../data/Panopea_generosa_scaff-70k_bowtie2-index.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 1807373\n",
      "    bwtLen: 1807374\n",
      "    sz: 451844\n",
      "    bwtSz: 451844\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 112961\n",
      "    offsSz: 451844\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 9414\n",
      "    numLines: 9414\n",
      "    ebwtTotLen: 602496\n",
      "    ebwtTotSz: 602496\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:01\n"
     ]
    }
   ],
   "source": [
    "# Bowtie cannot find the index that I created. \n",
    "# I'll try this: download Bowtie2 and try running through the same process:\n",
    "!/Users/shlaura3/anaconda/bin/bowtie2-build \\\n",
    "../data/Panopea_generosa_scaff-70k.fasta \\\n",
    "../data/Panopea_generosa_scaff-70k_bowtie2-index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u0001\u0000\u0000\u0000\r",
      "\u001b\u0000\u0006\u0000\u0000\u0000\u0002\u0000\u0000\u0000\u0004\u0000\u0000\u0000\r\n",
      "\u0000\u0000\u0000\u0014\u0000\u0000\u0000@i\u0001\u0000x\u0001\u00002\u0001\u0000\u0001\u0000\u0001\u0000?\u0017\u0001\u0000\u000es\u0001\u0000?\u001e",
      "\u0001\u0000W\u0001\u0000(b\u0001\u0000\u0013]\u0002\u0000N\u0001\u0000xA\u0001\u0000j\u0001\u0000b'\u0001\u0000X\u0001\u0000Hf\u0001\u00003\u0001\u0000}3\u0001\u0000\u0001\u0000l\u0002\u0000\u0000\u0000\u0000\u0000\u0000\u0013\u0000\u0000\u0000}\u0001\u0000s\u0000\u0000\u0000\u0013\u0000\u0000\u0000\u0001\u0000\u0001\u0000\u0000\u0013\u0000\u0000\u0000u\u0001\u0000\r\n",
      "\u0000\u0000\u0013\u0000\u0000\u0000\u0001\u0000\f",
      "\u0014\u0000\u0000\u0013\u0000\u0000\u0000\u0010\u0001\u0000\u0014\u0000\u0000\u0013\u0000\u0000\u0000c\u0001\u0000\u0014/\u0000\u0000\u0013\u0000\u0000\u0000\tc\u0001\u0000/\u0000\u0000\u0013\u0000\u0000\u0000^\u0001\u00003\u0000\u0000\u0013\u0000\u0000\u0000V\u0001\u0000;\u0000\u0000\u0013\u0000\u0000\u0000t7\u0001\u00008Z\u0000\u0000\u0013\u0000\u0000\u0000!6\u0001\u0000][\u0000\u0000\u0013\u0000\u0000\u00005\u0001\u0000[\u0000\u0000\u0013\u0000\u0000\u0000$5\u0001\u0000[\u0000\u0000\u0013\u0000\u0000\u00003\u0001\u0000\\\u0000\u0000\u0013\u0000\u0000\u00000\u0001\u0000^\u0000\u0000\u0013\u0000\u0000\u0000/\u0001\u0000L_\u0000\u0000\u0013\u0000\u0000\u0000_.\u0001\u0000r`\u0000\u0000\u0013\u0000\u0000\u0000c*\u0001\u0000c\u0000\u0000\u0013\u0000\u0000\u0000\u0012\u0001\u0000z\u0000\u0000\u0013\u0000\u0000\u0000\f",
      "\u0001\u0000\u0000\u0000\u0013\u0000\u0000\u0000c\u0000\u0000j\u0000\u0000\u0013\u0000\u0000\u0000p\u0000\u0000,\u0000\u0000\u0013\u0000\u0000\u0000O\u0000\u0000\u0000\u0000\u0013\u0000\u0000\u0000\u0000\u0000\u001f\u0000\u0000\u0013\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0013\u0000\u0000\u0000\u0000\u0000-\u0000\u0000\u0013\u0000\u0000\u0000\u000e\u0000\u0000\u0000\u0000\u0013\u0000\u0000\u0000#p\u0000\u0000\u0016\u0001\u0000\u0013\u0000\u0000\u0000`\u0000\u0000E%\u0001\u0000\u0013\u0000\u0000\u0000_\u0000\u0000h%\u0001\u0000\u0013\u0000\u0000\u0000\u0019N\u0000\u00006\u0001\u0000\u0013\u0000\u0000\u0000yG\u0000\u0000g<\u0001\u0000\u0013\u0000\u0000\u0000\u0018F\u0000\u0000<\u0001\u0000\u0013\u0000\u0000\u00009-\u0000\u0000$T\u0001\u0000\u0013\u0000\u0000\u0000\u001b\u0000\u00006e\u0001\u0000\u0013\u0000\u0000\u0000\u0011\u0000\u0000\u0019n\u0001\u0000\u0013\u0000\u0000\u0000>\u0011\u0000\u00009n\u0001\u0000\u0013\u0000\u0000\u0000\r",
      "\u0000\u0000\bq\u0001\u0000\u0013\u0000\u0000\u0000\u0000\u0000\u0000\u0000{}\u0001\u0000\u0012\u0000\u0000\u0000Q\t\u0001\u0000\u0001\u0000\u0012\u0000\u0000\u0000\u0000\u0000\u0001\u0000\u0012\u0000\u0000\u0000\u0000\u0000\u0001\u0000\u0012\u0000\u0000\u0000\u0000\u0000p\u0001\u0000\u0012\u0000\u0000\u0000\u0000\u0000\u0003\u0001\u0000\u0012\u0000\u0000\u0000\u0000\u0000E\u0001\u0000\u0012\u0000\u0000\u0000\u0000\u0000\u0001\u0000\u0012\u0000\u0000\u0000V\u0000\u0000w\u0001\u0000\u0012\u0000\u0000\u0000\u0000\u0000\u0007\u0002\u0000\u0012\u0000\u0000\u0000#\u0000\u0000\b\u0002\u0000\u0012\u0000\u0000\u0000\u0000\u0000}\r",
      "\u0002\u0000\u0012\u0000\u0000\u0000~\u0000\u0000\u0012\u0002\u0000\u0012\u0000\u0000\u0000\u0000\u0000\u0016\u0002\u0000\u0012\u0000\u0000\u0000\u0000\u0000\u0000\u001d",
      "\u0002\u0000\u0012\u0000\u0000\u0000\u0000\u0000k\u001f\u0002\u0000\u0012\u0000\u0000\u0000\u0000\u0000%\u0002\u0000\u0012\u0000\u0000\u0000Xz\u0000\u0000+\u0002\u0000\u0012\u0000\u0000\u0000x\u0000\u0000e,\u0002\u0000\u0012\u0000\u0000\u0000Gx\u0000\u0000,\u0002\u0000\u0012\u0000\u0000\u0000\u0015u\u0000\u0000D/\u0002\u0000\u0012\u0000\u0000\u0000in\u0000\u0000W5\u0002\u0000\u0012\u0000\u0000\u0000_\u0000\u0000aC\u0002\u0000\u0012\u0000\u0000\u0000^\u0000\u0000C\u0002\u0000\u0012\u0000\u0000\u0000]\u0000\u0000\u0018D\u0002\u0000\u0012\u0000\u0000\u0000@,\u0000\u0000s\u0002\u0000\u0012\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0002\u0000\u0011\u0000\u0000\u0000\u0001\u0000F\u0002\u0000\u0011\u0000\u0000\u0000S\u0001\u0000\u0002\u0000\u0011\u0000\u0000\u0000D\u0001\u00004\u0002\u0000\u0011\u0000\u0000\u0000SD\u0001\u0000T\u0002\u0000\u0011\u0000\u0000\u0000G6\u0001\u00005\u0006\u0003\u0000\u0011\u0000\u0000\u0000/\u0001\u0000\r\n",
      "\u0003\u0000\u0011\u0000\u0000\u0000P.\u0001\u0000\u000b",
      "\u0003\u0000\u0011\u0000\u0000\u0000\u0000\u0000+i\u0003\u0000\u0011\u0000\u0000\u0000\u0000\u0000\u0011y\u0003\u0000\u0011\u0000\u0000\u0000\u0000\u0000}\u0003\u0000\u0011\u0000\u0000\u0000\u0000\u0000\u0019\u0003\u0000\u0011\u0000\u0000\u0000E\u0000\u0000\u0003\u0000\u0011\u0000\u0000\u0000\u0000\u0000\u0003\u0000\u0011\u0000\u0000\u0000s\u0000\u0000{\u0003\u0000\u0011\u0000\u0000\u0000Rs\u0000\u0000\u0003\u0000\u0011\u0000\u0000\u0000\u001d",
      "s\u0000\u0000\u0003\u0000\u0011\u0000\u0000\u0000\u00041\u0000\u0000\u0004\u0004\u0000\u0011\u0000\u0000\u0000?/\u0000\u0000k\u0005\u0004\u0000\u0011\u0000\u0000\u0000\u0003\u0000\u00000\u0004\u0000\u0011\u0000\u0000\u0000\u0002\u0000\u00001\u0004\u0000\u0011\u0000\u0000\u0000\u0000\u0000\u0000\u00003\u0004\u0000\u0010\u0000\u0000\u0000\u001bf\u0001\u00003\u0004\u0000\u0010\u0000\u0000\u0000!d\u0001\u00004\u0004\u0000\u0010\u0000\u0000\u0000zc\u0001\u00004\u0004\u0000\u0010\u0000\u0000\u0000\u0015T\u0001\u0000C\u0004\u0000\u0010\u0000\u0000\u0000\u0015H\u0001\u0000EO\u0004\u0000\u0010\u0000\u0000\u0000\u0007\u0000\u0000\u0004\u0000\u0010\u0000\u0000\u0000\u0000\u0000p\u0004\u0000\u0010\u0000\u0000\u0000F\u0000\u0000\u0004\u0004\u0000\u0010\u0000\u0000\u0000 \u0000\u0000\u0004\u0000\u0010\u0000\u0000\u0000\u0000\u0000\u0004\u0000\u0010\u0000\u0000\u0000\u0000\u0000\u0004\u0000\u0010\u0000\u0000\u0000\u0012\u0000\u0000N\u0004\u0000\u0010\u0000\u0000\u0000\u0000\u0000f\u0004\u0000\u0010\u0000\u0000\u0000P\u0000\u0000d\u0004\u0000\u0010\u0000\u0000\u0000e\u0000\u0000\u0004\u0000\u0010\u0000\u0000\u0000et\u0000\u0000\u001c",
      "\u0005\u0000\u0010\u0000\u0000\u0000G\u0000\u0000I\u0005\u0000\u0010\u0000\u0000\u0000D>\u0000\u0000R\u0005\u0000\u0010\u0000\u0000\u0000/\u0000\u0000`\u0005\u0000\u0010\u0000\u0000\u0000\u001d",
      "\u0000\u0000Lr\u0005\u0000\u0010\u0000\u0000\u0000)\u0017\u0000\u0000x\u0005\u0000\u0010\u0000\u0000\u0000\u001e",
      "\u0014\u0000\u0000z\u0005\u0000\u0010\u0000\u0000\u0000\u0014\u0005\u0000\u0000\u0005\u0000\u0010\u0000\u0000\u0000\u0001\u0000\u0000o\u0005\u0000\u0010\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0005\u0000\u000f\u0000\u0000\u0000X\u0001\u0000\u0003\u0005\u0000\u000f\u0000\u0000\u0000V\u0001\u0000<\u0005\u0000\u000f\u0000\u0000\u0000S\u0001\u0000F\u0005\u0000\u000f\u0000\u0000\u0000MS\u0001\u0000q\u0005\u0000\u000f\u0000\u0000\u0000K\u0001\u0000\u0005\u0000\u000f\u0000\u0000\u0000C\u0001\u0000K\u0005\u0000\u000f\u0000\u0000\u00001B\u0001\u0000x\u0005\u0000\u000f\u0000\u0000\u0000\"A\u0001\u0000\u0005\u0000\u000f\u0000\u0000\u0000B@\u0001\u0000\u0005\u0000\u000f\u0000\u0000\u0000'<\u0001\u0000\u0005\u0000\u000f\u0000\u0000\u00009\u0001\u0000\u0005\u0000\u000f\u0000\u0000\u0000x9\u0001\u0000/\u0005\u0000\u000f\u0000\u0000\u0000#\u0001\u0000\u0005\u0000\u000f\u0000\u0000\u0000\u001b\u0001\u0000!\u0005\u0000\u000f\u0000\u0000\u0000+\u001b\u0001\u0000c\u0005\u0000\u000f\u0000\u0000\u0000\u0019\u0001\u0000\u0005\u0000\u000f\u0000\u0000\u0000;\u000f\u0001\u0000\u0005\u0000\u000f\u0000\u0000\u0000\u0000\u0001\u0000\u001b\u0005\u0000\u000f\u0000\u0000\u0000k\u0000\u0000?\u0005\u0000\u000f\u0000\u0000\u0000\u0000\u0000K\u0005\u0000\u000f\u0000\u0000\u0000&\u0000\u0000r\u0005\u0000\u000f\u0000\u0000\u0000\u0000\u0000\b\u0005\u0000\u000f\u0000\u0000\u0000\u0000\u0000\u0005\u0000\u000f\u0000\u0000\u0000\u0014\u0000\u00009\u0005\u0000\u000f\u0000\u0000\u0000\u0000\u0000\u0005\u0000\u000f\u0000\u0000\u0000\u0000\u0000l\b\u0006\u0000\u000f\u0000\u0000\u0000\u0003\u0000\u0000\u0018\u0006\u0000\u000f\u0000\u0000\u0000\u0000\u0000A\u001e",
      "\u0006\u0000\u000f\u0000\u0000\u0000Y@\u0000\u0000g\u0006\u0000\u000f\u0000\u0000\u0000\u0014\u0000\u0000\u0006\u0000\u000f\u0000\u0000\u0000D\u000f\u0000\u0000*\u0006\u0000\u000f\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0006\u0000\u000e\u0000\u0000\u0000\u0012&\u0001\u0000C\u0006\u0000\u000e\u0000\u0000\u0000\u0015#\u0001\u0000\u0006\u0000\u000e\u0000\u0000\u0000\"\u0001\u0000\u001f\u0006\u0000\u000e\u0000\u0000\u0000\u001b\u0001\u0000\u0006\u0000\u000e\u0000\u0000\u0000\t\u001b\u0001\u0000\u0006\u0000\u000e\u0000\u0000\u0000p\u001a\u0001\u0000q\u0006\u0000\u000e\u0000\u0000\u0000}\u0019\u0001\u0000\u0006\u0000\u000e\u0000\u0000\u0000\u0010\u0001\u0000\u0006\u0000\u000e\u0000\u0000\u0000F\u0002\u0001\u0000\u0006\u0000\u000e\u0000\u0000\u0000\u0000\u0000\u0000\u0007\u0000\u000e\u0000\u0000\u0000\u0000\u0000\u0002\u0007\u0000\u000e\u0000\u0000\u0000\u0000\u00009\u0007\u0007\u0000\u000e\u0000\u0000\u00001\u0000\u0000\\\u0007\u0007\u0000\u000e\u0000\u0000\u0000\u0000\u0000\u000e\b\u0007\u0000\u000e\u0000\u0000\u0000-\u0000\u0000\u000b",
      "\u0007\u0000\u000e\u0000\u0000\u0000\u0000\u0000\u000b",
      "\u0007\u0000\u000e\u0000\u0000\u0000\u0000\u0000\r",
      "\u0007\u0000\u000e\u0000\u0000\u0000*\u0000\u0000#\u0007\u0000\u000e\u0000\u0000\u0000\u0000\u0000\u0002$\u0007\u0000\u000e\u0000\u0000\u0000\u0000\u0000s-\u0007\u0000\u000e\u0000\u0000\u0000\u0000\u0000_F\u0007\u0000\u000e\u0000\u0000\u0000\u0000\u0000F\u0007\u0000\u000e\u0000\u0000\u0000V\u0000\u0000F\u0007\u0000\u000e\u0000\u0000\u0000\u0001\u0000\u0000\u000eI\u0007\u0000\u000e\u0000\u0000\u0000o\u0000\u0000[\u0007\u0000\u000e\u0000\u0000\u0000ZG\u0000\u0000Y\u0007\u0000\u000e\u0000\u0000\u0000\u0010F\u0000\u0000q\u0007\u0000\u000e\u0000\u0000\u0000@+\u0000\u0000\u000e\u0007\u0000\u000e\u0000\u0000\u0000%\u0000\u0000\u0007\u0000\u000e\u0000\u0000\u0000R$\u0000\u0000\u0007\u0000\u000e\u0000\u0000\u0000N\u0012\u0000\u0000)\u0007\u0000\u000e\u0000\u0000\u0000k\u000b",
      "\u0000\u0000~\u0007\u0000\u000e\u0000\u0000\u0000\u0001\u0000\u00002\u0007\u0000\u000e\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0007\u0000\r",
      "\u0000\u0000\u0000fj\u0001\u0000=\u0007\u0000\r",
      "\u0000\u0000\u0000d\u0001\u0000<\u0007\u0000\r",
      "\u0000\u0000\u00000c\u0001\u0000\u0007\u0000\r",
      "\u0000\u0000\u0000Y\u0001\u0000\u0007\u0000\r",
      "\u0000\u0000\u0000P\u0001\u0000\u0003\b\u0000\r",
      "\u0000\u0000\u0000J\u0001\u0000\t\b\u0000\r",
      "\u0000\u0000\u00000\u0001\u0000\"\b\u0000\r",
      "\u0000\u0000\u0000\u0019&\u0001\u0000^-\b\u0000\r",
      "\u0000\u0000\u0000.%\u0001\u0000\u0014.\b\u0000\r",
      "\u0000\u0000\u0000$\u0001\u0000U.\b\u0000\r",
      "\u0000\u0000\u0000!\u0001\u0000\u00151\b\u0000\r",
      "\u0000\u0000\u0000\f",
      "\u0018\u0001\u0000V:\b\u0000\r",
      "\u0000\u0000\u0000m\u0011\u0001\u0000(@\b\u0000\r",
      "\u0000\u0000\u00005\u0000\u0000_e\b\u0000\r",
      "\u0000\u0000\u0000\u0012\u0000\u0000\b\u0000\r",
      "\u0000\u0000\u0000\u0000\u0000n\b\u0000\r",
      "\u0000\u0000\u0000\u0000\u0000\b\u0000\r",
      "\u0000\u0000\u0000\u001d",
      "\u0000\u0000)\b\u0000\r",
      "\u0000\u0000\u0000p\u0000\u0000\b\u0000\r",
      "\u0000\u0000\u00001o\u0000\u0000\b\u0000\r",
      "\u0000\u0000\u0000l\u0000\u0000t\b\u0000\r",
      "\u0000\u0000\u0000l\u0000\u0000\b\u0000\r",
      "\u0000\u0000\u0000k\u0000\u0000-\b\u0000\r",
      "\u0000\u0000\u0000j\u0000\u0000\b\u0000\r",
      "\u0000\u0000\u0000j\u0000\u0000\u001e",
      "\b\u0000\r",
      "\u0000\u0000\u0000\u001c",
      "j\u0000\u0000m\b\u0000\r",
      "\u0000\u0000\u0000h\u0000\u0000G\b\u0000\r",
      "\u0000\u0000\u0000\u0010h\u0000\u0000\b\u0000\r",
      "\u0000\u0000\u0000`\u0000\u0000\b\u0000\r",
      "\u0000\u0000\u0000O]\u0000\u00005\b\u0000\r",
      "\u0000\u0000\u0000\\\\\u0000\u0000a\b\u0000\r",
      "\u0000\u0000\u0000\t;\u0000\u0000\u0013\u0010\t\u0000\r",
      "\u0000\u0000\u0000\u0000\u0000\u0000\u0000yJ\t\u0000\f",
      "\u0000\u0000\u0000@<\u0001\u0000O\t\u0000\f",
      "\u0000\u0000\u00007\u0001\u0000R\t\u0000\f",
      "\u0000\u0000\u00007\u0001\u00005S\t\u0000\f",
      "\u0000\u0000\u00005\u0001\u0000T\t\u0000\f",
      "\u0000\u0000\u0000/\u0001\u0000iZ\t\u0000\f",
      "\u0000\u0000\u0000#\u0001\u0000f\t\u0000\f",
      "\u0000\u0000\u0000\"\u0001\u0000\u0015g\t\u0000\f",
      "\u0000\u0000\u0000\b\u0017\u0001\u0000#r\t\u0000\f",
      "\u0000\u0000\u0000h\u0000\u0000\t\u0000\f",
      "\u0000\u0000\u0000S\u0000\u0000U\t\u0000\f",
      "\u0000\u0000\u0000S\u0000\u0000\t\u0000\f",
      "\u0000\u0000\u0000\u0000\u0000\b\t\u0000\f",
      "\u0000\u0000\u0000 \u0000\u0000@\t\u0000\f",
      "\u0000\u0000\u0000B\u0000\u0000\t\u0000\f",
      "\u0000\u0000\u0000\u0000\u0000\u0007\t\u0000\f",
      "\u0000\u0000\u0000l\u0000\u0000\t\u0000\f",
      "\u0000\u0000\u00000\u0000\u0000B\t\u0000\f",
      "\u0000\u0000\u0000\u0000\u0000[\u0002\r\n",
      "\u0000\f",
      "\u0000\u0000\u0000f\u0000\u0000J\u001b\r\n",
      "\u0000\f",
      "\u0000\u0000\u0000D\u0000\u0000\u0010<\r\n",
      "\u0000\f",
      "\u0000\u0000\u00005\u000b",
      "\u0000\u0000ut\r\n",
      "\u0000\f",
      "\u0000\u0000\u0000\u0000\u0000\u0000\u0000X\r\n",
      "\u0000\u000b",
      "\u0000\u0000\u0000%\u0001\u0000\r\n",
      "\u0000\u000b",
      "\u0000\u0000\u0000\u0001\u0000\r\n"
     ]
    }
   ],
   "source": [
    "! head ../data/Panopea_generosa_scaff-70k_bowtie2-index.rev.1.bt2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bowtie 2 version 2.2.8 by Ben Langmead (langmea@cs.jhu.edu, www.cs.jhu.edu/~langmea)\r\n",
      "Usage: \r\n",
      "  bowtie2 [options]* -x <bt2-idx> {-1 <m1> -2 <m2> | -U <r>} [-S <sam>]\r\n",
      "\r\n",
      "  <bt2-idx>  Index filename prefix (minus trailing .X.bt2).\r\n",
      "             NOTE: Bowtie 1 and Bowtie 2 indexes are not compatible.\r\n",
      "  <m1>       Files with #1 mates, paired with files in <m2>.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <m2>       Files with #2 mates, paired with files in <m1>.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <r>        Files with unpaired reads.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <sam>      File for SAM output (default: stdout)\r\n",
      "\r\n",
      "  <m1>, <m2>, <r> can be comma-separated lists (no whitespace) and can be\r\n",
      "  specified many times.  E.g. '-U file1.fq,file2.fq -U file3.fq'.\r\n",
      "\r\n",
      "Options (defaults in parentheses):\r\n",
      "\r\n",
      " Input:\r\n",
      "  -q                 query input files are FASTQ .fq/.fastq (default)\r\n",
      "  --qseq             query input files are in Illumina's qseq format\r\n",
      "  -f                 query input files are (multi-)FASTA .fa/.mfa\r\n",
      "  -r                 query input files are raw one-sequence-per-line\r\n",
      "  -c                 <m1>, <m2>, <r> are sequences themselves, not files\r\n",
      "  -s/--skip <int>    skip the first <int> reads/pairs in the input (none)\r\n",
      "  -u/--upto <int>    stop after first <int> reads/pairs (no limit)\r\n",
      "  -5/--trim5 <int>   trim <int> bases from 5'/left end of reads (0)\r\n",
      "  -3/--trim3 <int>   trim <int> bases from 3'/right end of reads (0)\r\n",
      "  --phred33          qualities are Phred+33 (default)\r\n",
      "  --phred64          qualities are Phred+64\r\n",
      "  --int-quals        qualities encoded as space-delimited integers\r\n",
      "\r\n",
      " Presets:                 Same as:\r\n",
      "  For --end-to-end:\r\n",
      "   --very-fast            -D 5 -R 1 -N 0 -L 22 -i S,0,2.50\r\n",
      "   --fast                 -D 10 -R 2 -N 0 -L 22 -i S,0,2.50\r\n",
      "   --sensitive            -D 15 -R 2 -N 0 -L 22 -i S,1,1.15 (default)\r\n",
      "   --very-sensitive       -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\r\n",
      "\r\n",
      "  For --local:\r\n",
      "   --very-fast-local      -D 5 -R 1 -N 0 -L 25 -i S,1,2.00\r\n",
      "   --fast-local           -D 10 -R 2 -N 0 -L 22 -i S,1,1.75\r\n",
      "   --sensitive-local      -D 15 -R 2 -N 0 -L 20 -i S,1,0.75 (default)\r\n",
      "   --very-sensitive-local -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\r\n",
      "\r\n",
      " Alignment:\r\n",
      "  -N <int>           max # mismatches in seed alignment; can be 0 or 1 (0)\r\n",
      "  -L <int>           length of seed substrings; must be >3, <32 (22)\r\n",
      "  -i <func>          interval between seed substrings w/r/t read len (S,1,1.15)\r\n",
      "  --n-ceil <func>    func for max # non-A/C/G/Ts permitted in aln (L,0,0.15)\r\n",
      "  --dpad <int>       include <int> extra ref chars on sides of DP table (15)\r\n",
      "  --gbar <int>       disallow gaps within <int> nucs of read extremes (4)\r\n",
      "  --ignore-quals     treat all quality values as 30 on Phred scale (off)\r\n",
      "  --nofw             do not align forward (original) version of read (off)\r\n",
      "  --norc             do not align reverse-complement version of read (off)\r\n",
      "  --no-1mm-upfront   do not allow 1 mismatch alignments before attempting to\r\n",
      "                     scan for the optimal seeded alignments\r\n",
      "  --end-to-end       entire read must align; no clipping (on)\r\n",
      "   OR\r\n",
      "  --local            local alignment; ends might be soft clipped (off)\r\n",
      "\r\n",
      " Scoring:\r\n",
      "  --ma <int>         match bonus (0 for --end-to-end, 2 for --local) \r\n",
      "  --mp <int>         max penalty for mismatch; lower qual = lower penalty (6)\r\n",
      "  --np <int>         penalty for non-A/C/G/Ts in read/ref (1)\r\n",
      "  --rdg <int>,<int>  read gap open, extend penalties (5,3)\r\n",
      "  --rfg <int>,<int>  reference gap open, extend penalties (5,3)\r\n",
      "  --score-min <func> min acceptable alignment score w/r/t read length\r\n",
      "                     (G,20,8 for local, L,-0.6,-0.6 for end-to-end)\r\n",
      "\r\n",
      " Reporting:\r\n",
      "  (default)          look for multiple alignments, report best, with MAPQ\r\n",
      "   OR\r\n",
      "  -k <int>           report up to <int> alns per read; MAPQ not meaningful\r\n",
      "   OR\r\n",
      "  -a/--all           report all alignments; very slow, MAPQ not meaningful\r\n",
      "\r\n",
      " Effort:\r\n",
      "  -D <int>           give up extending after <int> failed extends in a row (15)\r\n",
      "  -R <int>           for reads w/ repetitive seeds, try <int> sets of seeds (2)\r\n",
      "\r\n",
      " Paired-end:\r\n",
      "  -I/--minins <int>  minimum fragment length (0)\r\n",
      "  -X/--maxins <int>  maximum fragment length (500)\r\n",
      "  --fr/--rf/--ff     -1, -2 mates align fw/rev, rev/fw, fw/fw (--fr)\r\n",
      "  --no-mixed         suppress unpaired alignments for paired reads\r\n",
      "  --no-discordant    suppress discordant alignments for paired reads\r\n",
      "  --no-dovetail      not concordant when mates extend past each other\r\n",
      "  --no-contain       not concordant when one mate alignment contains other\r\n",
      "  --no-overlap       not concordant when mates overlap at all\r\n",
      "\r\n",
      " Output:\r\n",
      "  -t/--time          print wall-clock time taken by search phases\r\n",
      "  --un <path>           write unpaired reads that didn't align to <path>\r\n",
      "  --al <path>           write unpaired reads that aligned at least once to <path>\r\n",
      "  --un-conc <path>      write pairs that didn't align concordantly to <path>\r\n",
      "  --al-conc <path>      write pairs that aligned concordantly at least once to <path>\r\n",
      "  (Note: for --un, --al, --un-conc, or --al-conc, add '-gz' to the option name, e.g.\r\n",
      "  --un-gz <path>, to gzip compress output, or add '-bz2' to bzip2 compress output.)\r\n",
      "  --quiet            print nothing to stderr except serious errors\r\n",
      "  --met-file <path>  send metrics to file at <path> (off)\r\n",
      "  --met-stderr       send metrics to stderr (off)\r\n",
      "  --met <int>        report internal counters & metrics every <int> secs (1)\r\n",
      "  --no-unal          suppress SAM records for unaligned reads\r\n",
      "  --no-head          suppress header lines, i.e. lines starting with @\r\n",
      "  --no-sq            suppress @SQ header lines\r\n",
      "  --rg-id <text>     set read group id, reflected in @RG line and RG:Z: opt field\r\n",
      "  --rg <text>        add <text> (\"lab:value\") to @RG line of SAM header.\r\n",
      "                     Note: @RG line only printed when --rg-id is set.\r\n",
      "  --omit-sec-seq     put '*' in SEQ and QUAL fields for secondary alignments.\r\n",
      "\r\n",
      " Performance:\r\n",
      "  -p/--threads <int> number of alignment threads to launch (1)\r\n",
      "  --reorder          force SAM output order to match order of input reads\r\n",
      "  --mm               use memory-mapped I/O for index; many 'bowtie's can share\r\n",
      "\r\n",
      " Other:\r\n",
      "  --qc-filter        filter out reads that are bad according to QSEQ filter\r\n",
      "  --seed <int>       seed for random number generator (0)\r\n",
      "  --non-deterministic seed rand. gen. arbitrarily instead of using read attributes\r\n",
      "  --version          print version information and quit\r\n",
      "  -h/--help          print this usage message\r\n"
     ]
    }
   ],
   "source": [
    "! /Users/shlaura3/anaconda/bin/bowtie2 -help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/shlaura3/Documents/SAFS/FISH_546_Bioinformatics/546-Bioinformatics/2016-10_Geo-Ann-Project/data'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Flags\t1\r\n",
      "Reverse flags\t5\r\n",
      "Colorspace\t0\r\n",
      "2.0-compatible\t1\r\n",
      "SA-Sample\t1 in 16\r\n",
      "FTab-Chars\t10\r\n",
      "Sequence-1\tscaffold3071 37.0\t92480\r\n",
      "Sequence-2\tscaffold4463 35.0\t96504\r\n",
      "Sequence-3\tscaffold5354 34.4\t78472\r\n",
      "Sequence-4\tscaffold9504 36.1\t106643\r\n",
      "Sequence-5\tscaffold10970 36.6\t103072\r\n",
      "Sequence-6\tscaffold11875 36.6\t71487\r\n",
      "Sequence-7\tscaffold15463 36.3\t94990\r\n",
      "Sequence-8\tscaffold18558 34.9\t73279\r\n",
      "Sequence-9\tscaffold19489 36.3\t88023\r\n",
      "Sequence-10\tscaffold20302 34.0\t90664\r\n",
      "Sequence-11\tscaffold26337 37.5\t154899\r\n",
      "Sequence-12\tscaffold26960 37.7\t118606\r\n",
      "Sequence-13\tscaffold27692 37.9\t82296\r\n",
      "Sequence-14\tscaffold30278 35.7\t92860\r\n",
      "Sequence-15\tscaffold31723 35.8\t75618\r\n",
      "Sequence-16\tscaffold32578 36.5\t88248\r\n",
      "Sequence-17\tscaffold34940 36.6\t91720\r\n",
      "Sequence-18\tscaffold45727 36.4\t106547\r\n",
      "Sequence-19\tscaffold59644 35.9\t78717\r\n",
      "Sequence-20\tscaffold71773 33.8\t103920\r\n"
     ]
    }
   ],
   "source": [
    "! bowtie2-inspect --summary ../data/Panopea_generosa_scaff-70k_bowtie2-index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "103213920 reads; of these:\n",
      "  103213920 (100.00%) were paired; of these:\n",
      "    102865097 (99.66%) aligned concordantly 0 times\n",
      "    261264 (0.25%) aligned concordantly exactly 1 time\n",
      "    87559 (0.08%) aligned concordantly >1 times\n",
      "    ----\n",
      "    102865097 pairs aligned concordantly 0 times; of these:\n",
      "      14413 (0.01%) aligned discordantly 1 time\n",
      "    ----\n",
      "    102850684 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      205701368 mates make up the pairs; of these:\n",
      "        205379458 (99.84%) aligned 0 times\n",
      "        205853 (0.10%) aligned exactly 1 time\n",
      "        116057 (0.06%) aligned >1 times\n",
      "0.51% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "# Now, use bowtie to first map female gonad paired-end reads to scaffold index\n",
    "# FYI this took about an hour\n",
    "! /Users/shlaura3/anaconda/bin/bowtie2 \\\n",
    "-x ../data/Panopea_generosa_scaff-70k_bowtie2-index \\\n",
    "-1 Geo_Pool_F_GGCTAC_L006_R1_001_val_2.fq \\\n",
    "-2 Geo_Pool_F_GGCTAC_L006_R2_001_val_1.fq \\\n",
    "-S Panopea_generosascaff-70k_gonadF_bowtie.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "! mv ../data/Panopea_generosascaff-70k_gonadF_bowtie.sam ../analyses/Panopea_generosascaff-70k_gonadF_bowtie.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2016-11-3_Geo70k-scaff-transcrip-sequences.tabular\r\n",
      "GeoTrans-PhelDiffExp_blasted_sorted\r\n",
      "P-Generosa_IGV.xml\r\n",
      "Panopea_generosascaff-70k_gonadF_bowtie.sam\r\n",
      "Phel_DEGlist.tab\r\n",
      "\u001b[34mRepeatMasker\u001b[m\u001b[m\r\n",
      "SeaStarDiffExp.R\r\n",
      "pgenerosa-scaff70-miRNA.tab\r\n",
      "pgenerosa-transcrv3-blastn-scaff70k-01.gff.fai\r\n",
      "pgenerosa-transcrv3-blastn-scaff70k-01.tab\r\n",
      "pgenerosa-transcrv3-blastn-scaff70k-01.tab.fai\r\n"
     ]
    }
   ],
   "source": [
    "!ls ../analyses/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@HD\tVN:1.0\tSO:unsorted\r\n",
      "@SQ\tSN:scaffold3071\tLN:92480\r\n",
      "@SQ\tSN:scaffold4463\tLN:96504\r\n",
      "@SQ\tSN:scaffold5354\tLN:78472\r\n",
      "@SQ\tSN:scaffold9504\tLN:106643\r\n",
      "@SQ\tSN:scaffold10970\tLN:103072\r\n",
      "@SQ\tSN:scaffold11875\tLN:71487\r\n",
      "@SQ\tSN:scaffold15463\tLN:94990\r\n",
      "@SQ\tSN:scaffold18558\tLN:73279\r\n",
      "@SQ\tSN:scaffold19489\tLN:88023\r\n"
     ]
    }
   ],
   "source": [
    "! head ../analyses/Panopea_generosascaff-70k_gonadF_bowtie.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Converting from .sam to .bam format\n",
    "! samtools view -bS -o \\\n",
    "../analyses/Panopea_generosascaff-70k_gonadF_bowtie.bam \\\n",
    "../analyses/Panopea_generosascaff-70k_gonadF_bowtie.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2016-11-3_Geo70k-scaff-transcrip-sequences.tabular\r\n",
      "GeoTrans-PhelDiffExp_blasted_sorted\r\n",
      "P-Generosa_IGV.xml\r\n",
      "Panopea_generosascaff-70k_gonadF_bowtie.bam\r\n",
      "Panopea_generosascaff-70k_gonadF_bowtie.sam\r\n",
      "Phel_DEGlist.tab\r\n",
      "\u001b[34mRepeatMasker\u001b[m\u001b[m/\r\n",
      "SeaStarDiffExp.R\r\n",
      "pgenerosa-scaff70-miRNA.tab\r\n",
      "pgenerosa-transcrv3-blastn-scaff70k-01.gff.fai\r\n",
      "pgenerosa-transcrv3-blastn-scaff70k-01.tab\r\n",
      "pgenerosa-transcrv3-blastn-scaff70k-01.tab.fai\r\n"
     ]
    }
   ],
   "source": [
    "ls ../analyses/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 91 files...\r\n"
     ]
    }
   ],
   "source": [
    "! samtools sort \\\n",
    "-o ../analyses/Panopea_generosascaff-70k_gonadF_bowtie-sorted.bam \\\n",
    "-T TEMP_gonadF_bowtie-sorted \\\n",
    "../analyses/Panopea_generosascaff-70k_gonadF_bowtie.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/shlaura3/Documents/SAFS/FISH_546_Bioinformatics/546-Bioinformatics/2016-10_Geo-Ann-Project/analyses\n"
     ]
    }
   ],
   "source": [
    "cd ../analyses/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2016-11-3_Geo70k-scaff-transcrip-sequences.tabular\r\n",
      "GeoTrans-PhelDiffExp_blasted_sorted\r\n",
      "P-Generosa_IGV.xml\r\n",
      "Panopea_generosascaff-70k_gonadF_bowtie-sorted.bam\r\n",
      "Panopea_generosascaff-70k_gonadF_bowtie.bam\r\n",
      "Panopea_generosascaff-70k_gonadF_bowtie.sam\r\n",
      "Phel_DEGlist.tab\r\n",
      "\u001b[34mRepeatMasker\u001b[m\u001b[m/\r\n",
      "SeaStarDiffExp.R\r\n",
      "pgenerosa-scaff70-miRNA.tab\r\n",
      "pgenerosa-transcrv3-blastn-scaff70k-01.gff.fai\r\n",
      "pgenerosa-transcrv3-blastn-scaff70k-01.tab\r\n",
      "pgenerosa-transcrv3-blastn-scaff70k-01.tab.fai\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools: unrecognized option `--help'\r\n",
      "\r\n",
      "Usage: samtools mpileup [options] in1.bam [in2.bam [...]]\r\n",
      "\r\n",
      "Input options:\r\n",
      "  -6, --illumina1.3+      quality is in the Illumina-1.3+ encoding\r\n",
      "  -A, --count-orphans     do not discard anomalous read pairs\r\n",
      "  -b, --bam-list FILE     list of input BAM filenames, one per line\r\n",
      "  -B, --no-BAQ            disable BAQ (per-Base Alignment Quality)\r\n",
      "  -C, --adjust-MQ INT     adjust mapping quality; recommended:50, disable:0 [0]\r\n",
      "  -d, --max-depth INT     max per-file depth; avoids excessive memory usage [250]\r\n",
      "  -E, --redo-BAQ          recalculate BAQ on the fly, ignore existing BQs\r\n",
      "  -f, --fasta-ref FILE    faidx indexed reference sequence file\r\n",
      "  -G, --exclude-RG FILE   exclude read groups listed in FILE\r\n",
      "  -l, --positions FILE    skip unlisted positions (chr pos) or regions (BED)\r\n",
      "  -q, --min-MQ INT        skip alignments with mapQ smaller than INT [0]\r\n",
      "  -Q, --min-BQ INT        skip bases with baseQ/BAQ smaller than INT [13]\r\n",
      "  -r, --region REG        region in which pileup is generated\r\n",
      "  -R, --ignore-RG         ignore RG tags (one BAM = one sample)\r\n",
      "  --rf, --incl-flags STR|INT  required flags: skip reads with mask bits unset []\r\n",
      "  --ff, --excl-flags STR|INT  filter flags: skip reads with mask bits set\r\n",
      "                                            [UNMAP,SECONDARY,QCFAIL,DUP]\r\n",
      "  -x, --ignore-overlaps   disable read-pair overlap detection\r\n",
      "\r\n",
      "Output options:\r\n",
      "  -o, --output FILE       write output to FILE [standard output]\r\n",
      "  -g, --BCF               generate genotype likelihoods in BCF format\r\n",
      "  -v, --VCF               generate genotype likelihoods in VCF format\r\n",
      "\r\n",
      "Output options for mpileup format (without -g/-v):\r\n",
      "  -O, --output-BP         output base positions on reads\r\n",
      "  -s, --output-MQ         output mapping quality\r\n",
      "\r\n",
      "Output options for genotype likelihoods (when -g/-v is used):\r\n",
      "  -t, --output-tags LIST  optional tags to output:\r\n",
      "               DP,AD,ADF,ADR,SP,INFO/AD,INFO/ADF,INFO/ADR []\r\n",
      "  -u, --uncompressed      generate uncompressed VCF/BCF output\r\n",
      "\r\n",
      "SNP/INDEL genotype likelihoods options (effective with -g/-v):\r\n",
      "  -e, --ext-prob INT      Phred-scaled gap extension seq error probability [20]\r\n",
      "  -F, --gap-frac FLOAT    minimum fraction of gapped reads [0.002]\r\n",
      "  -h, --tandem-qual INT   coefficient for homopolymer errors [100]\r\n",
      "  -I, --skip-indels       do not perform indel calling\r\n",
      "  -L, --max-idepth INT    maximum per-file depth for INDEL calling [250]\r\n",
      "  -m, --min-ireads INT    minimum number gapped reads for indel candidates [1]\r\n",
      "  -o, --open-prob INT     Phred-scaled gap open seq error probability [40]\r\n",
      "  -p, --per-sample-mF     apply -m and -F per-sample for increased sensitivity\r\n",
      "  -P, --platforms STR     comma separated list of platforms for indels [all]\r\n",
      "      --input-fmt-option OPT[=VAL]\r\n",
      "               Specify a single input file format option in the form\r\n",
      "               of OPTION or OPTION=VALUE\r\n",
      "      --reference FILE\r\n",
      "               Reference sequence FASTA FILE [null]\r\n",
      "\r\n",
      "Notes: Assuming diploid individuals.\r\n"
     ]
    }
   ],
   "source": [
    "! samtools mpileup --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[E::hts_open_format] fail to open file '../analyses/Panopea_generosascaff-70k_gonadF_bowtie-sorted.bam'\r\n",
      "[mpileup] failed to open ../analyses/Panopea_generosascaff-70k_gonadF_bowtie-sorted.bam: No such file or directory\r\n"
     ]
    }
   ],
   "source": [
    "! samtools mpileup -u -t DP \\\n",
    "-f ../data/Panopea_generosa_scaff-70k.fasta \\\n",
    "../analyses/Panopea_generosascaff-70k_gonadF_bowtie-sorted.bam \\\n",
    "-o ../analyses/Panopea_generosascaff-70k_gonadF_mpileup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geo-70k-scaff-annotations-merged-table.tabular\r\n",
      "Geo-v3-join-uniprot-all0916-condensed.tab\r\n",
      "Geoduck-transcriptome-v3.fa.zip\r\n",
      "Geoduck-transcriptome-v3.fa_db.nhr\r\n",
      "Geoduck-transcriptome-v3.fa_db.nin\r\n",
      "Geoduck-transcriptome-v3.fa_db.nsq\r\n",
      "Panopea_generosa_No-Line-Breaks.scafSeq.fai\r\n",
      "Panopea_generosa_ScafSeq.genome\r\n",
      "Panopea_generosa_scaff-70k.fasta\r\n",
      "Panopea_generosa_scaff-70k.fasta.fai\r\n",
      "Panopea_generosa_scaff-70k.scafSeq\r\n",
      "Panopea_generosa_scaff-70k.scafSeq.fai\r\n",
      "Panopea_generosa_scaff-70k_bowtie2-index.1.bt2\r\n",
      "Panopea_generosa_scaff-70k_bowtie2-index.2.bt2\r\n",
      "Panopea_generosa_scaff-70k_bowtie2-index.3.bt2\r\n",
      "Panopea_generosa_scaff-70k_bowtie2-index.4.bt2\r\n",
      "Panopea_generosa_scaff-70k_bowtie2-index.rev.1.bt2\r\n",
      "Panopea_generosa_scaff-70k_bowtie2-index.rev.2.bt2\r\n",
      "Panopea_generosa_scaff-70k_db.nhr\r\n",
      "Panopea_generosa_scaff-70k_db.nin\r\n",
      "Panopea_generosa_scaff-70k_db.nsq\r\n",
      "Phel_countdata.txt\r\n",
      "Phel_transcriptome.fasta\r\n",
      "hairpin.fa\r\n"
     ]
    }
   ],
   "source": [
    "ls ../data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
