#### This directory contains _Panopea generosa_ genome annotation files. They are numerically labeled, corresponding to the Jupyter Notebook recording the steps taken to create these files.  Below is a list of the files categorized by Jupyter Notebook #.
**Notice that files from Notebook 07 are housed in Owl.**


| Jupyter Notebook                                                                                                                                                                                                                                     | Associated Results Files                                                                                                                                                                                       | File Description                                                                                                                                      |
|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------|
| [01 Extracting >70k Scaffolds.ipynb](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/Jupyter-Notebooks/01%20Extracting%20%3E70k%20Scaffolds.ipynb)                                                           | [01-Panopea_generosa_scaff-70k.fasta](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/results/01-Panopea_generosa_scaff-70k.fasta)                                     | Fasta sequence file with only scaffolds >70k bp                                                                                                       |
|                                                                                                                                                                                                                                                      | [01-Panopea_generosa_scaff-70k.fasta.fai](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/results/01-Panopea_generosa_scaff-70k.fasta.fai)                             | Index file for .fasta file (necessary for visualizing in IGV)                                                                                         |
| [02 Blasting Transcriptome vs. Scaffolds.ipynb](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/Jupyter-Notebooks/02%20Blasting%20Transcriptome%20vs.%20Scaffolds.ipynb)                                     | [02-pgenerosa-transcrv3-blastn-scaff70k-01.gff](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/results/02-pgenerosa-transcrv3-blastn-scaff70k-01.gff)                 | Where gonad transcriptome aligns with the genome scaffolds; output from Blast program                                                                 |
| [03 Merging blast results with transcriptome GO details.ipynb](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/Jupyter-Notebooks/03%20Merging%20blast%20results%20with%20transcriptome%20GO%20details.ipynb) | [03-Geo-70k-scaff-annotations-merged-table.tabular](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/results/03-Geo-70k-scaff-annotations-merged-table.tabular)         | Large, data-rich file that provides genome & scaffold indexing, Uniprot protein data, GO terms, & bp sequences                                        |
| [04 Finding Transposable Elements.ipynb](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/Jupyter-Notebooks/04%20Finding%20Transposable%20Elements.ipynb)                                                     | [04-Panopea_generosa_scaff-70k.scafSeq.out.refined.gff](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/results/04-Panopea_generosa_scaff-70k.scafSeq.out.refined.gff) | Indexed .gff file showing where transposable elements are expected; output from RepeatMasker program.                                                 |
| [05 Finding candidate methylation sites.ipynb](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/Jupyter-Notebooks/05%20Finding%20candidate%20methylation%20sites.ipynb)                                       | [05-Panopea_generosa_Scaff-70k-CpG.gff](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/results/05-Panopea_generosa_Scaff-70k-CpG.gff)                                 | Indexed .gff file with candidate methylation sites (simply, locations of CG); output of Galaxy's EMBOSS fuzznuc tool.                                 |
| [06 Finding candidate miRNA.ipynb](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/Jupyter-Notebooks/06%20Finding%20candidate%20miRNA.ipynb)                                                                 | [06-pgenerosa-scaff70-miRNA.gff](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/results/06-pgenerosa-scaff70-miRNA.gff)                                               | Indexed .gff file showing where miRNA is expected; output from blasting the downloadable miRBase hairpin database against genome scaffolds.           |
|                                                                                                                                                                                                                                                      | [06-pgenerosa-scaff70-miRNA.tab](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/results/06-pgenerosa-scaff70-miRNA.tab)                                               | Same content as above, but .tab format (prior to converting to .gff)                                                                                  |
| [07 Bowtie2_map_reads.ipynb](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/Jupyter-Notebooks/07%20Bowtie2_map_reads.ipynb)                                                                                 | **On Owl** [Panopea_generosascaff-70k_gonadF_bowtie-sorted.bam](http://owl.fish.washington.edu/generosa/GeoAnn/)                                                                                               |  Indexed .bam files where RNASeq reads from **female** gonad tissue align with genome scaffolds. File is large (15GB); output  from `bowtie2` aligner |
|                                                                                                                                                                                                                                                      | [Panopea_generosascaff-70k_gonadF_bowtie-sorted.bam.bai](http://owl.fish.washington.edu/generosa/GeoAnn/)                                                                                                      | Index file for above .bam file; must be in the same directory for IGV                                                                                 |
|                                                                                                                                                                                                                                                      | [Panopea_generosascaff-70k_gonadM_bowtie-sorted.bam](http://owl.fish.washington.edu/generosa/GeoAnn/)                                                                                                          | Indexed .bam files where RNASeq reads from **male** gonad tissue align with genome scaffolds. File is large (16GB); output,from `bowtie2` aligner     |
|                                                                                                                                                                                                                                                      | [Panopea_generosascaff-70k_gonadM_bowtie-sorted.bam.bai](http://owl.fish.washington.edu/generosa/GeoAnn/)                                                                                                      | Index file for above .bam file; must be in the same directory for IGV                                                                                 |
|                                                                                                                                                                                                                                                      | [md5.txt](http://owl.fish.washington.edu/generosa/GeoAnn/)                                                                                                                                                     | MD5 file provides md5 codes for .bam files originally generated on the "Roadrunner" computer in Roberts lab, and on the Owl server.                   |
| [08 Calling_SNPS_mpileup.ipynb](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/Jupyter-Notebooks/08%20Calling_SNPS_mpileup.ipynb)                                                                           | [08-Pg_70k_gonadF_SNPs.vcf](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/results/08-Pg_70k_gonadF_SNPs.vcf)                                                         | Indexed files with SNP locations in the **female** gonads; output from `samtools | bcftools`                                                          |
|                                                                                                                                                                                                                                                      | [08-Pg_70k_gonadM_SNPs.vcf](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/results/08-Pg_70k_gonadM_SNPs.vcf)                                                         | Indexed files with SNP locations in the **female** gonads; output from `samtools | bcftools`                                                          |
| [09 Visualizing with IGV .ipynb](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/Jupyter-Notebooks/09%20Visualizing%20with%20IGV%20.ipynb)                                                                   | [09-Pg_70k_tracks.xml](https://github.com/laurahspencer/546-Bioinformatics/blob/master/2016-10_Geo-Ann-Project/results/09-Pg_70k_tracks.xml)                                                                   | .xml file with IGV Session data; open this file through the IGV software, and tracks will load via URLs. No need to download files.                   |


## For a glimpse of the results, check out the below IGV images of a few scaffolds:

![All scaffolds view]()

![Scaff 4463]()

![Scaff 32578]()

## Discussion 

In approaching this project I did not have a research question that needed answering, rather my goal was to characterize genetic markers for _P. generosa_. As such, the results section does not include a data analysis.  Next steps are likely the following: 
  * Determine germline DNA methylation estimate in scaffolds via CpG O/E analysis.
  * Identify candidate lncRNA regions
  * Experiment with other databases to identify miRNA, and compare results to those in this project; database used was miRBase, and I'm interested to know how the results differ with other dtabases.
  * Experiment with other programs to identify transposable elements, and compare results to those in this project; program used in this project was RepeatMasker.
  * Develop a thorough QC and screening workflow; as my comfort with these data types increases, I will be able to confidently edit out low e-value hits, and identiy errors/anomalies.
